File: Arista Networks/18.4-sum.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        /*
        Approach:
        Sort the array, fix two non-duplicate anchors, then use two pointers for
        the remaining pair. long long is used for sums so large int inputs do not
        overflow while comparing to target.

        Complexity: O(n^3) time and O(log n) to O(n) sort stack space.
        */
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        int n = (int)nums.size();
        for (int i = 0; i < n - 3; ++i) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            for (int j = i + 1; j < n - 2; ++j) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;
                int left = j + 1, right = n - 1;
                while (left < right) {
                    long long sum = (long long)nums[i] + nums[j] + nums[left] + nums[right];
                    if (sum == target) {
                        ans.push_back({nums[i], nums[j], nums[left], nums[right]});
                        int lv = nums[left], rv = nums[right];
                        while (left < right && nums[left] == lv) ++left;
                        while (left < right && nums[right] == rv) --right;
                    } else if (sum < target) {
                        ++left;
                    } else {
                        --right;
                    }
                }
            }
        }
        return ans;
    }
};

File: companies/ServiceNow/18.4-sum.java
import java.util.*;

/**
 * Algorithm:
 * Sort the array, fix two unique indices, and solve the remaining 2Sum with
 * two pointers. Skip duplicates at every layer.
 *
 * Java data structures:
 * Arrays.sort orders the input; ArrayList stores quadruplets. long is used for
 * the sum to avoid Java int overflow on large values.
 *
 * Complexity:
 * Time O(n^3), extra space O(1) excluding output.
 */
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();
        int n = nums.length;
        for (int i = 0; i < n - 3; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            for (int j = i + 1; j < n - 2; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) {
                    continue;
                }
                int left = j + 1;
                int right = n - 1;
                while (left < right) {
                    long total = (long) nums[i] + nums[j] + nums[left] + nums[right];
                    if (total == target) {
                        ans.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));
                        left++;
                        right--;
                        while (left < right && nums[left] == nums[left - 1]) {
                            left++;
                        }
                        while (left < right && nums[right] == nums[right + 1]) {
                            right--;
                        }
                    } else if (total < target) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
        }
        return ans;
    }
}

